libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/ptx/parallel_synchronization_and_communication_instructions_mbarrier.h

```diff
@@ -0,0 +1,392 @@
+ *   https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier
+ *
+ */
+
+// 9.7.12.15.9. Parallel Synchronization and Communication Instructions: mbarrier.init
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-init
+
+// 9.7.12.15.10. Parallel Synchronization and Communication Instructions: mbarrier.inval
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-inval
+
+// 9.7.12.15.11. Parallel Synchronization and Communication Instructions: mbarrier.expect_tx
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-expect-tx
+
+// 9.7.12.15.12. Parallel Synchronization and Communication Instructions: mbarrier.complete_tx
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-complete-tx
+
+// 9.7.12.15.13. Parallel Synchronization and Communication Instructions: mbarrier.arrive
+// https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier-arrive
+
+// mbarrier.arrive{.sem}{.scope}{.shared{::cta}}.b64           state, [addr]{, count};
+// mbarrier.arrive{.sem}{.scope}{.shared::cluster}.b64         _, [addr] {,count}
+// mbarrier.arrive.expect_tx{.sem}{.scope}{.shared{::cta}}.b64 state, [addr], txCount;
+// mbarrier.arrive.expect_tx{.sem}{.scope}{.shared::cluster}.b64   _, [addr], txCount;
+// mbarrier.arrive.noComplete{.sem}{.cta}{.shared{::cta}}.b64  state, [addr], count;
+//
+// .sem   = { .release }
+// .scope = { .cta, .cluster }
+
+/*
+// mbarrier.arrive.shared.b64 state, [addr]; // 1.  PTX ISA 70, SM_80
+template <typename=void>
+__device__ static inline uint64_t mbarrier_arrive(
+  cuda::ptx::sem_release_t,
+  cuda::ptx::scope_cta_t,
+  cuda::ptx::space_shared_t,
+  uint64_t* addr);
+*/
+#if __cccl_ptx_isa >= 700
+template <typename=void>
+_LIBCUDACXX_DEVICE static inline _CUDA_VSTD::uint64_t mbarrier_arrive(
+  sem_release_t,
+  scope_cta_t,
+  space_shared_t,
+  _CUDA_VSTD::uint64_t* __addr)
+{
+  // __sem == sem_release (due to parameter type constraint)
+  // __scope == scope_cta (due to parameter type constraint)
+  // __space == space_shared (due to parameter type constraint)
+
+  _CUDA_VSTD::uint64_t __state;
+
+  NV_IF_ELSE_TARGET(NV_PROVIDES_SM_80,(
+    asm (
+      "mbarrier.arrive.shared.b64 %0, [%1]; // 1. "
+      : "=l"(__state)
+      : "r"(__as_ptr_smem(__addr))
+      : "memory"
+    );
+    return __state;
+  ),(
+    // Unsupported architectures will have a linker error with a semi-decent error message
+    __device__ _CUDA_VSTD::uint64_t ___cuda_vstd_uint64_t__cuda_ptx_mbarrier_arrive_is_not_supported_before_SM_80__();
```

Comment:
suggestion: we can make the error a bit more readable if we hoist the declaration:
```cpp
extern "C" _CUDA_VSTD::uint64_t ___cuda_vstd_uint64_t__cuda_mbarrier_arrive_is_not_supported_before_SM_80__();
```
Leads to:
```bash
 Unresolved extern function 'cuda_vstd_uint64_t__cuda_ptx_mbarrier_arrive_is_not_supported_before_SM_80__'
```

vs

```bash
Unresolved extern function '_Z76cuda_vstd_uint64_t__cuda_ptx_mbarrier_arrive_is_not_supported_before_SM_80__v'
```

Reply:
That is a great suggestion! I noticed that the name mangling could make the function unreadable, but had not considered using extern C!

I had to wrestle a bit to get it to work, because `extern "C"` declarations appear to be disallowed inside the function body. In any case, I pushed the change. 

Thank you very much again for the original suggestion and this one makes it even better.